fclient/performjoin.go

```diff
@@ -0,0 +1,238 @@
+package fclient
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/sirupsen/logrus"
+)
+
+type PerformJoinInput struct {
+	UserID     *spec.UserID
+	RoomID     string
+	ServerName spec.ServerName
+	Content    map[string]interface{}
+	Unsigned   map[string]interface{}
+
+	PrivateKey ed25519.PrivateKey
+	KeyID      gomatrixserverlib.KeyID
+	KeyRing    *gomatrixserverlib.KeyRing
+
+	EventProvider func(
+		ctx context.Context, fedClient FederationClient,
+		keyRing gomatrixserverlib.JSONVerifier, origin, server spec.ServerName,
+	) gomatrixserverlib.EventProvider
+}
+
+type PerformJoinCallbacks struct {
+	FederationFailure func(serverName spec.ServerName)
+	FederationSuccess func(serverName spec.ServerName)
+}
+
+func PerformJoin(
+	ctx context.Context,
+	fedClient FederationClient,
+	input PerformJoinInput,
+	callbacks PerformJoinCallbacks,
+) (*gomatrixserverlib.HeaderedEvent, gomatrixserverlib.StateResponse, error) {
```

Comment:
I don't think this function signature is ideal, because it doesn't really handle errors particularly elegantly. The return args are otherwise good (the join event and the state response, decoupled from the format it is over federation). Error wise, there's a few things we care about:
 - Server is unreachable. Transient error.
 - The server's response is invalid. Permanent error.
 - There's a problem on our side e.g unknown room version or failed to sign the event. Permanent error.

In addition, we care if the server was reachable and we received an error, for blacklist purposes. I would probably define a general `FederationError` in GMSL (not fclient as it isn't http specific) which looks something like:
```go
type FederationError struct {
    ServerName spec.ServerName
    Transient bool
    Reachable bool
    Err error
}

func (e FederationError) Error() string {
    return fmt.Sprintf("FederationError(t=%v,r=%v): %w", e.Transient, e.Reachable, e.Err)
}
``` 

This would mean:
 - Server is unreachable. Transient=true, Reachable=false
 - The server's response is invalid. Transient=false, Reachable=true
 - Failure to sign the event on our side. Transient=false, Reachable=true
 - Failure to talk to our own DB e.g to load keys in the KeyRing. Transient=true, Reachable=true